jdaviz/configs/default/plugins/markers/markers.py

```diff
@@ -89,6 +93,17 @@ def _create_viewer_callbacks(self, viewer):
     def _on_viewer_added(self, msg):
         self._create_viewer_callbacks(self.app.get_viewer_by_id(msg.viewer_id))
 
+    def _on_refdata_change(self, msg):
+        viewer_mark = self._get_mark(msg.viewer)
+        # we act from the previous state instead of the original since the original WCS may have
+        # been deleted.  If this causes a buildup in errors, we could first try to access the
+        # original or cache its WCS in the table so we always have access.
+        prev_x, prev_y = viewer_mark.x, viewer_mark.y
+        prev_wcs = msg.old.coords
+        new_wcs = msg.data.coords
+        new_x, new_y = new_wcs.world_to_pixel(prev_wcs.pixel_to_world(prev_x*u.pix, prev_y*u.pix))
+        viewer_mark.x, viewer_mark.y = new_x, new_y

```

Comment:
On second thought, this probably needs more logic when changing link-type.... 🤔

Reply:
ok, I am pleasantly surprised that I'm wrong and this seems to just work, although I'm not sure why 😆

Reply:
I agree. As it currently stands, when I click on stuff while pixel-linked, it reports that it knows I was looking at the second (non-reference image) and reports the pixels for that.

![Screenshot 2023-11-03 161739](https://github.com/bmorris3/jdaviz/assets/2090236/df737c82-c2d1-43b4-88ac-0a2f115e4cf9)

Then I change link type to WCS, and I thought it would try to follow where I marked the point on the second image while pixel-linked. But I am really not sure what is the intended behavior here when you designed this.

![Screenshot 2023-11-03 161522](https://github.com/bmorris3/jdaviz/assets/2090236/5ce827f2-4f5a-4477-bb51-6abac3101ff7)

Should we clear all the markers when changing link type?

Reply:
Can you show a video or a script to reproduce?  This is what I'm seeing (I need to investigate why it is showing blank for link type instead of saying its pixel-linked, but that's different from what we're discussing here)


https://github.com/bmorris3/jdaviz/assets/877591/80f5f390-8890-47ea-8c2d-69b4bd54e52b

> Should we clear all the markers when changing link type?

I'd rather try to make it work.  The table should remain unchanged, but we should have all the information needed to rotate the marks as needed.

Reply:
> Can you show a video or a script to reproduce?

Load both the images in `ImvizExample.ipynb` but do not link them by WCS yet. At this point, they are linked by pixel. Blink to A. Then using your Markers plugin, on A, mark some points. Here was what I marked:

![Screenshot 2023-11-03 171349](https://github.com/bmorris3/jdaviz/assets/2090236/10aaf5ba-2006-463a-a08b-a182ee8915b7)

Now blink to B. Mark something else.

![Screenshot 2023-11-03 171524](https://github.com/bmorris3/jdaviz/assets/2090236/58adab2d-a015-49f4-b65b-24aa404c126d)

You would see something like the table I posted above at https://github.com/bmorris3/jdaviz/pull/15#discussion_r1382162359 . Now toggle "keep active" on and link them by WCS. I would expect the 5th point that I mark on B to move but it didn't, as shown in https://github.com/bmorris3/jdaviz/pull/15#discussion_r1382162359 .

Reply:
Weirdly enough, I do not see your problem with blank link type:

![Screenshot 2023-11-03 171900](https://github.com/bmorris3/jdaviz/assets/2090236/c05c7c0f-8534-4c1d-9ef7-8c6d92020c38)

![Screenshot 2023-11-03 171947](https://github.com/bmorris3/jdaviz/assets/2090236/21ba94b9-53a9-4fab-b28e-de1343a01c80)

Reply:
I think this is actually acting as I'd expect (given the current implementation) as it is translating from the original reference to the new reference.  Should we instead translate based on what was considered the active/top layer at the time?  Maybe 🤔  but that will also require making sure we have access to that WCS in case that layer has been deleted.  Or perhaps we clear marks that reference a deleted layer?

Reply:
If this behavior is expected, we need to rethink how we represent the info that can get misleading when link type or reference layer changes, but maybe that is out of scope here?

Reply:
I'll take a stab at tracking the original WCS and see if that behavior feels more natural/intuitive.

Reply:
In short, I think we have to decide: does the mark "stick" to the reference data or the active layer on which it was created when the link type is changed?  If we can't decide or that is too difficult to implement, I think we can defer that as out-of-scope for now since this is supposed to explicitly add support for rotation (but if we can decide - now would be a nice time to try to make that work).  I think we can all agree that for a change in rotation, we'd expect the markers to move, and there is no difference between the two scenarios since we're guaranteed to be WCS linked at that point.

For example, here is a hacky implementation of sticking to the "active" layer (although at the end you can see that it doesn't remember for the reverse operation, which is where this choice will get a bit complicated to implement).


https://github.com/bmorris3/jdaviz/assets/877591/575c2e0f-25ef-4f2c-97ab-548f7951bd6c

Reply:
What do you think about this option?

* If pixel linked, only remember X/Y.
* If WCS linked, only remember SkyCoord.
* When change linking type, wipe them off since original intention is lost.
* When rotate, since it is always WCS linked, re-render on display pixel using SkyCoord.

Reply:
The intention of the markers plugin is to act as a log of the mouseover information, and its definitely useful to have access to all of the information.  If we want to say that only pixel information is relevant when pixel-linked, then we should remove that from the mouseover display as well (and vice-versa).

That said, we could stop drawing the actual markers (with a warning) - I just don't want to bail to that option yet since it feels like just giving up and not thinking it through.  I think we can either choose what it sticks to or give the user the option and they will quickly learn and that is less obtrusive than having to clear on a change to link type.

Reply:
As far as I can tell, the only time we can not correctly preserve marker information on changes in link type is when markers are defined on data without WCS, since switching to WCS-linking will drop that layer, and its pixel-to-world conversion isn't defined. This is the only time we truly need to drop markers, so I think that's the only case we can treat differently.

Reply:
Log is not useful if the intention is lost. So, what exactly are we logging that is meaningful to the user in all these different options?

Reply:
Either pixel coordinates and their corresponding data layer, or world coordinates – whichever we have access to.

Reply:
> Log is not useful if the intention is lost. 

I'm not 100% sure what you mean, but if you mean that coordinate logs aren't useful when the layer is rotated or is not visible in the viewer, I don't agree.

Reply:
I was responding to "its definitely useful to have access to all of the information." I do not think it is useful to keep pixel info when linked by WCS now given all the additional complexity.

Reply:
I think the table should always show the information that was in the coordinate display at the time that the marker was added, with all the context necessary to make sense of that information.

It sounds like the consensus is to have the markers stick to the "active" data layer (rather than the reference), with any case where that isn't intended (semi-transparency, etc) controllable by the cycler in the mouseover display.  That also means we can probably hide the reference data and link type section from the table since it won't be needed for that logic.  I also think this means that the pixel information in the table would both remain relevant and not be ambiguous - it is the pixel of the state data layer for that marker, and that remains fixed by any rotation or change in reference.

I'll see if I can get a general solution here, but may need to make the WCS->pixel case a follow-up effort if it gets too complicated.